File: hw/final/src/ffbs/Gibbs.R
gibbs <- function(init, update, B=2000, burn=100, printFreq=0) {

  out <- as.list(1:B)
  out[[1]] <- init

  for (i in 2:(B+burn)) {
    out[[i]] <- update(out[[i-1]])
    if (printFreq > 0 && i%%printFreq==0) cat("\rProgress: ",i,"/",B+burn)
  }

  tail(out,B)
}

rig <- function(shp,rate) 1 / rgamma(1,shp,rate)

File: hw/final/src/scala/ffbs/src/main/scala/Gibbs.scala
package ffbs.mcmc

object Gibbs {

  import breeze.stats.distributions.{Gamma,Gaussian}
  import scala.util.Random.{nextGaussian=>randn, nextDouble=>randu}

  def rig(shp: Double, rate: Double) = 1.0 / (new Gamma(shp, 1.0/rate)).sample
  def tnorm(mu: Double, sig2: Double, lower: Double, upper: Double) = {

  }

  def timer[R](block: => R) = {
    val t0 = System.nanoTime()
    val result = block
    val t1 = System.nanoTime()
    println("Elapsed time: " + (t1 - t0) / 1E9 + "s")
    result
  }

  trait State { 
    def update(): State
    def sample(B:Int, burn:Int, printEvery:Int=0) = {
      def loop(S:List[State], i:Int): List[State] = {
        if (printEvery > 0 && i % printEvery == 0) 
          print("\rProgress: " + i +"/"+ (B+burn) + "\t")

        if (i < B + burn) {
          val newState = if (i <= burn) 
            List(S.head.update)
          else
            S.head.update :: S

          loop(newState, i+1)
        } else S
      }
      loop(List(this),0).asInstanceOf[List[this.type]]
    }
  }

}
